docs/rules-cookbook.md

```diff
@@ -0,0 +1,519 @@
+  matches:
+    all:
+    - paths:
+        begins_with:
+          - '/images/'
+    - paths:
+        ends_with:
+          - '.jpg'
+```
+
+Match paths that either match one of a list of exact strings or a regex:
+
+```
+  matches:
+    any:
+      - paths:
+          regex:
+            - '/path/to/matching/directory/.+'
+      - paths:
+          exact:
+            - '/a/very/specific/request/path/'
+            - '/another/path/'
+            - '/the/last/exact/path/'
+```
+
+### 1.2  Query parameter matching
+
+Match a request based on a specific value for a parameter:
+
+```
+  matches:
+    any:
+      - query:
+          parameter: my_parameter
+          exact:
+            - required_value
+```
+
+Match requests on multiple parameters, case-insensitive on the second parameter which can begin with either `foo` or `bar`:
+
+```
+  matches:
+    all:
+      - query:
+          parameter: my_parameter
+          exact:
+            - required_value
+      - query:
+          parameter: other_parameter
+          ignore_case: true
+          begins_with:
+             - foo
+             - bar
+```
+
+### 1.3  Negative matching
+
+Match any paths that are NOT the paths `/public/` or `/html/public/`:
+
+```
+  matches:
+    all:
+      - paths:
+          not: True
+          begins_with:
+            - '/public/'
+            - '/html/public/'
+```
+
+Match any path that does not END with `/public/`:
+```
+  matches:
+    all:
+      - paths:
+          not: True
+          ends_with:
+            - '/public/'
+```
+
+### 1.4  Combining different matching methods
+
+Match paths that begin with either `/secret/` or `/topsecret/` EXCEPT the subdirectories `public`:
+
+```
+  matches:
+    all:
+      - paths:
+          begins_with:
+            - '/secret/'
+            - '/topsecret/'
+      - paths:
+          not: True
+          begins_with:
+            - '/secret/public/'
+            - '/topsecret/public/'
+```
+
+As above, but using `ends_with` for negative matching instead:
+```
+  matches:
+    all:
+      - paths:
+          begins_with:
+            - '/secret/'
+            - '/topsecret/'
+      - paths:
+          not: True
+          ends_with:
+            - '/public/'
+```
+### 1.5 `domain_default` matching
+
+The `matches` directive can only be present in a rule that does *NOT* have the `domain_default` property set, for obvious reasons. A rule that has `domain_default: True` will match any requests to that domain that are not matched by a rule for the same domain that does explicit matching.
+
+Complete rule example:
+
+```
+- description: www.domain.local - domain default
+  domains:
+    - www.mydomain.com
+  domain_default: True
+  actions:
+    backend:
+      origin: 'https://default-origin.localdomain'
+    header_southbound:
+      - set:
+        field: 'Host'
+        value: 'default-origin.localdomain'
+      - set:
+        field: 'X-Forwarded-Host'
+        value: 'www.mydomain.com'
+```
+
+## 2. Routing
```

Comment:
Is it more or less readable/clear if we put all actions (Routing, Rewriting, Redirecting) under an actions heading?

Instead of the current:
```
2. Routing
3. Rewriting
4. Redirecting
```
To have:
```
2. Actions
  2.1. Routing
  2.2. Rewriting
  2.3. Redirecting
```

Reply:
The idea with the chapter names was to provide keys that are maybe a little more in line with the user's intentions and needs rather than reflecting the syntax structure, but on the other hand grouping them under the Actions heading doesn't really take away from that so let's go with your version.

Reply:
Done in 481da38a6081c78fd2592432c26b2ce73933ebee